MDEV-33710 Implement UUID_TIMESTAMP() function

varundeepsaini opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-33710*

## Description

Adds `UUID_TIMESTAMP(uuid)` function to extract timestamps from UUIDv1 and UUIDv7 values. Returns `NULL` for other versions (e.g., v4) or invalid input.

## Release Notes

New function: `UUID_TIMESTAMP(uuid)` - Returns the generation timestamp from UUIDv1/v7 as `TIMESTAMP(6)`.

## How can this PR be tested?

```bash
./mysql-test/mtr --suite=type_uuid func_uuid_timestamp
```

## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the `main` branch.*
- [ ] *This is a bug fix.*

## PR quality check
- [x] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [x] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -16,6 +16,10 @@
 #define MYSQL_SERVER
 #include "mariadb.h"
 #include "item_uuidfunc.h"
+#include <myisampack.h>
```
myisam is some storage engine, i.e. a plugin, ideally. It's not correct to use its library in general.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +33,53 @@ String *Item_func_sys_guid::val_str(String *str)
+  }
+  else if (version == 1)
+  {
+    ulonglong uuid_ts= ((ulonglong)(mi_uint2korr(buf + 6) & 0x0FFF) << 48) |
```
There are functions without mi_ prefix

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +33,53 @@ String *Item_func_sys_guid::val_str(String *str)
+  my_time_t seconds;
+  ulong microseconds;
+
+  if (version == 7)
```
Please explain the format for both in the comments.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +33,53 @@ String *Item_func_sys_guid::val_str(String *str)
+{
+  Type_std_attributes::set(
+    Type_temporal_attributes_not_fixed_dec(MAX_DATETIME_WIDTH,
+                                           TIME_SECOND_PART_DIGITS, false),
```
Why do you think it should be TIME_SECOND_PART_DIGITS?
Why unsigned_flag is false?

varundeepsaini replied:
UUIDv1 stores 7 hence i went with the max mariadb supports i.e 6

also for why the unsigned flag is false, i just did what other functions which use datetime also do

FooBarrior commented:
I have some concern about the tests... it's not easy to validate the UUID values used there. 
One approach I can think of is:
1. `set timestamp` to a fixed value. This will affect NOW() output
2. See that value returned by UUID_TIMESTAMP(UUID*()) matches our value

We can test a couple of random values, plus corner cases: min and max value for timestamp variable, and for UUID

varundeepsaini commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +33,53 @@ String *Item_func_sys_guid::val_str(String *str)
+}
+
+
+bool Item_func_uuid_timestamp::val_native(THD *thd, Native *to)
```
val_time is for duration right if im not mistaken ? 
like 2 mintes, 4 hour, 1 year etc, and not 12 Jan 2026

varundeepsaini commented:
i think the test failing is a timezone issue (exact 5.30 HRS), ill fix that

FooBarrior commented on mysys/my_uuid.c:
```diff
@@ -224,3 +224,61 @@ void my_uuid_end()
+
+  if (version == 1)
+  {
+    /* UUIDv1: reconstruct 60-bit timestamp from time_low, time_mid, time_hi */
```
Move the format documentation here. Explain what lo, mid and hi are.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +29,32 @@ String *Item_func_sys_guid::val_str(String *str)
+{
+  Type_handler_uuid_new::Fbt_null uuid(args[0]);
+  if (uuid.is_null())
+    return (null_value= true);
```
Every `return`  contains this `null_value` assignment. This means it makes sense to extract a function:

Make a private funciton which does all the conversion. 
In this public to_native, call that function and assign null_value.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +29,32 @@ String *Item_func_sys_guid::val_str(String *str)
+
+  my_time_t seconds;
+  ulong usec;
+  const uchar *buf= (const uchar *) uuid.to_lex_cstring().str;
```
No functions really need uchar (except the newly added one, which we can change), also uuid is natively stored in char, so why make a conversion?

varundeepsaini replied:
the mi_ prefix functions use uchar (they do cast it themselves tho, i can remove it from here but again, the cast is still there)

varundeepsaini replied:
https://github.com/varundeepsaini/server/blob/32e6f8ff2eb54f029e9fced9b17b780e7b7bd6d9/include/myisampack.h#L47-L51

FooBarrior commented on plugin/type_uuid/mysql-test/type_uuid/func_uuid_timestamp.test:
```diff
@@ -0,0 +1,51 @@
+SELECT UUID_TIMESTAMP('1ee81f00-a8a8-11ee-8000-000000000000') IS NOT NULL AS has_timestamp;
+
+# Test dynamic UUID generation - timestamp should match NOW()
+SELECT TIMESTAMPDIFF(SECOND, UUID_TIMESTAMP(UUID()), NOW()) AS v1_diff_seconds;
```
This test is (sort of, implicitly) indeterministic: each time it will check against the new NOW value. So if it will ever fail (because of the bug), noone will be able to reproduce it.

Fix NOW() with `set timestamp` statement.

FooBarrior replied:
Then using it vs NOW etc would mean test would fail sporadically. Can you show the example?

FooBarrior replied:
Besides we had no timestamp extraction before, so what do you mean

varundeepsaini replied:
i grepped NOW() / CURDATE() and found these tests 

https://github.com/varundeepsaini/server/blob/6a3fbfdb2d048a0a1b5762f8a56df35ba6986cf0/mysql-test/suite/innodb/t/update_time.test#L20-L25

https://github.com/varundeepsaini/server/blob/36a2a185fe18d31a644da46cfabd9757a379280c/mysql-test/suite/galera/t/MW-292.test#L76-L78

hence did something like that

FooBarrior commented on mysys/my_uuid.c:
```diff
@@ -224,3 +224,61 @@ void my_uuid_end()
+  if (version == 7)
+  {
+    /* UUIDv7: bytes 0-5 are Unix timestamp in milliseconds (big-endian) */
+    ts= mi_uint6korr(uuid);
```
mi_ functions are from a plugin, which is not correct to refer, this breaks the dependency order. 
use `uint6korr`, without prefix

varundeepsaini replied:
yeah

FooBarrior commented on mysys/my_uuid.c:
```diff
@@ -224,3 +224,61 @@ void my_uuid_end()
+    Timestamp is 100-nanosecond intervals since 1582-10-15
+
+  UUIDv7 format (RFC 9562):
+    Bytes 0-5:  Unix timestamp in milliseconds (48 bits, big-endian)
```
clarify which endianness is used in the format

FooBarrior reviewed (changes requested):
Thank you @varundeepsaini for the update, and besides such aquick one!
The PR moves in the right direction. Still some work to be done.

ParadoxV5 commented on mysys/my_uuid.c:
```diff
@@ -224,3 +224,66 @@ void my_uuid_end()
+  }
+
+  if (version == 1)
+  {
```
Beginner question: Would a `switch` block work here?

ParadoxV5 commented on mysys/my_uuid.c:
```diff
@@ -224,3 +224,61 @@ void my_uuid_end()
+    Bytes 6-7:  version (4 bits) + sub-millisecond precision (12 bits)
+*/
+
+my_bool my_uuid_extract_ts(const uchar *uuid, my_time_t *seconds, ulong *usec)
```
We might as well use `bool` here since there are only two result types.

Note for both `int` and `bool`: I hear that the convention is `1`/`true` = error (which one might consider “timestamp invalid” to be one) and `0`/`false` = no error.

ParadoxV5 reviewed (dismissed):
just passing by

FooBarrior commented on mysys/my_uuid.c:
```diff
@@ -224,3 +225,66 @@ void my_uuid_end()
+    @retval false  Success
+    @retval true   UUID version doesn't contain timestamp or timestamp invalid
+
+  UUIDv1 format (RFC 4122, big-endian):
```
If you cite RFC, also specify the section

FooBarrior commented on mysys/my_uuid.c:
```diff
@@ -224,3 +225,66 @@ void my_uuid_end()
+*/
+bool my_uuid_extract_ts(const char *uuid, my_time_t *seconds, ulong *usec)
+{
+  uint version= (uchar) uuid[6] >> 4;
```
just `char version`

varundeepsaini commented:
Generated the UUIDs from https://www.uuidgenerator.net/
and the min max UUIDs by hand

FooBarrior commented on mysys/my_uuid.c:
```diff
@@ -224,3 +225,66 @@ void my_uuid_end()
+      UUIDv1: reconstruct 60-bit timestamp from three fields:
+        - time_low  (bytes 0-3): bits 0-31 of timestamp
+        - time_mid  (bytes 4-5): bits 32-47 of timestamp
+        - time_hi   (bytes 6-7): bits 48-59 of timestamp (masked, 4 bits are version)
```
It would be either time_low/time_high, or time_lo/time_hi, but not the mix.
Let's use the RFC 9562 notion: time_high

gkodinov reviewed (approved):
This is a preliminary review. 
Can you please consider squashing all of the commits into a single one and have it have a commit message that complies with CODING_STANDARDS.md? 

Otherwise I see that the review is in full swing. Please keep working with the current reviewers until their concerns are addressed.

FooBarrior commented on plugin/type_uuid/mysql-test/type_uuid/func_uuid_timestamp.result:
```diff
@@ -0,0 +1,88 @@
+SELECT 'c28c09a0-e593-11f0-8de9-0242ac120002' AS uuid,
+UUID_TIMESTAMP('c28c09a0-e593-11f0-8de9-0242ac120002') AS ts;
+uuid	ts
+c28c09a0-e593-11f0-8de9-0242ac120002	2025-12-30 15:25:23.539600
```
If .001 is all the precision we can get out of UUID, then the type should be TIMESTAMP(3), not TIMESTAMP(6).

FooBarrior commented on plugin/type_uuid/plugin.cc:
```diff
@@ -180,16 +180,33 @@ class Create_func_uuid_v7 : public Create_func_arg0
   plugin_descriptor_function_uuid_v4(&Create_func_uuid_v4::s_singleton),
-  plugin_descriptor_function_uuid_v7(&Create_func_uuid_v7::s_singleton);
+  plugin_descriptor_function_uuid_v7(&Create_func_uuid_v7::s_singleton),
+  plugin_descriptor_function_uuid_timestamp(&Create_func_uuid_timestamp::s_singleton);
```
missing indent.

FooBarrior replied:
... or not. It's only looking so in the github diff. nvm.

FooBarrior commented on plugin/type_uuid/mysql-test/type_uuid/func_uuid_timestamp.test:
```diff
@@ -0,0 +1,66 @@
+# MDEV-33710: UUID_TIMESTAMP() extracts timestamp from UUIDv1 and UUIDv7
+
+SET time_zone='+00:00';
+
```
Add test: 
```
CREATE OR REPLACE TABLE t1 AS
  SELECT UUID_TIMESTAMP(UUID()) AS tsv1, 
         UUID_TIMESTAMP(UUID_V7()) as tsv7;
SHOW COLUMNS FROM t1;
DROP TABLE t1;
```

FooBarrior reviewed (changes requested):
Overall looks finally good to me, asked for a minor semantic fix and a test to output type.

According to RFC 9562, UUIDv7 is indeed limited with 3-digit timestamp precision. 
>unix_ts_ms:
>  48-bit big-endian unsigned number of the Unix Epoch timestamp **in milliseconds** as per [Section 6.1](https://www.rfc-editor.org/rfc/rfc9562#timestamp_considerations). Occupies bits 0 through 47 (octets 0-5).

After those things fixed, it's ok to push.
Please squash the commits.
The title should start from MDEV-XXXX
In commit message, add details about SQL functions added, their types. new c/c++ classes and functions. You can also mention and cite RFC, upon your will.

varundeepsaini committed:
MDEV-33710 Added UUID_TIMESTAMP() for UUIDv1 and UUIDv7

Added UUID_TIMESTAMP(expr) to the type_uuid plugin.

The function accepts values of the UUID type and UUID strings and
returns TIMESTAMP(6) NULL.

UUID_TIMESTAMP() returns NULL for UUID values before the Unix epoch,
for UUID versions without embedded time, and for invalid UUID input.

For UUIDv1, it rebuilds the 60-bit timestamp from the time_low,
time_mid, and time_high fields. Per RFC 9562 Section 5.1, this
timestamp is stored as 100-nanosecond intervals since 1582-10-15,
so the function converts it to Unix time.

For UUIDv7, it reads the 48-bit Unix Epoch timestamp in milliseconds
from the first 6 bytes. Per RFC 9562 Section 5.7 Figure 11, this is
the unix_ts_ms field, so the function converts milliseconds to the
MariaDB timestamp value.

Added my_uuid_extract_ts() in mysys/my_uuid.c and declared it in
include/my_sys.h. UUID_TIMESTAMP() calls this function to decode
timestamps from UUIDv1 and UUIDv7 values.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -29,3 +29,40 @@ String *Item_func_sys_guid::val_str(String *str)
+  */
+  Type_std_attributes::set(
+    Type_temporal_attributes_not_fixed_dec(MAX_DATETIME_WIDTH,
+                                           3, false),
```
Move 3 to constexpr, like UUID_TIMESTAMP_PRECISION

FooBarrior replied:
you may also want to cite the RFC in the comment to that constant's definition, highlighting the fact of millisecond precision

FooBarrior commented:
Please rebase on latest main as well.

FooBarrior commented on plugin/type_uuid/mysql-test/type_uuid/func_uuid_timestamp.result:
```diff
@@ -0,0 +1,89 @@
+# Return type metadata
+#
+CREATE OR REPLACE TABLE t1 AS
+  SELECT UUID_TIMESTAMP(UUID()) AS tsv1,
```
The test is faliing on buildbot, because the result is different from what would be outputted. Better run it with --record to get a .result file, rather that guessing, or using AI.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -17,6 +17,13 @@
 #include "item_uuidfunc.h"
 
+/*
+  RFC 9562 Section 5.7 defines the UUIDv7 timestamp field in milliseconds,
```
The solution is both for v1 and v7, and so the comment should mention both and cite the RFC regarding both of them.

FooBarrior reviewed (changes requested):
Turns out, UUIDv1 is 100-nanosec precision according to RFC. Though we don't support  that much: 100-ns is 7 digits, while mariadb maximum is 6 yet. 

I've checked the older commits and you have actually had the UUIDv1 results microsecond-precise. Why didn't you oppose my statement @varundeepsaini?

Hopefully, you don't have to revert the work done, just fix the constant and the results. Make sure we see the microsecond precision is in the result file (basically just as before).

We can't make a type dynamically deduced: UUID_TIMESTAMP accepts any string, so its resulting type should be well-defined (and it's defined by fix_lenght_and_dec), and as such, it's better to be a minimal precision that can hold the maximally-precised supported value.

varundeepsaini commented:
@FooBarrior 
My bad, I should have pointed that out earlier. This PR was inactive for a long time, and when I picked it up again that detail slipped my mind.

I’ll change it back to TIMESTAMP(6) and update the results.

FooBarrior commented:
Just please keep the value extracted into the constant.

FooBarrior reviewed (changes requested):
Please remove the AI-generated commit message, including sgned-off-by. Replace it with something thet will help others understand your code.

varundeepsaini commented:
@FooBarrior i think i need the sign off for the cla

FooBarrior commented on plugin/type_uuid/item_uuidfunc.cc:
```diff
@@ -17,6 +17,14 @@
 
+/*
+  UUIDv1 stores 100-nanosecond ticks (RFC 9562 Section 5.1) and UUIDv7 stores
+  Unix time in milliseconds (RFC 9562 Section 5.7). MariaDB supports up to
```
Still waiting on the RFC citation.

FooBarrior reviewed (changes requested):
> @FooBarrior i think i need the sign off for the cla

Never heard of this. Do you have a link? @gkodinov FYI

The commit message review follows, and now as it's human-written (as it seems) , it requires more work:

> Moved UUID time parsing to my_uuid_extract_ts() 

Moved from where? I though you implemented it.

> This keeps the UUID parsing logic in one shared place 

Is there any other UUID parsing logc??

> and makes the plugin code easier to read.

how?

> The UUIDv1 path rebuilds the timestamp from the UUID time fields and
converts 100-nanosecond ticks

The word "ticks" tickles in this context. Can be easily confused with clock ticks, which UUID1 also has. I'd prefer the RFC terms.

> to Unix time

converts from what

> The UUIDv7 path reads the first 48 bits as Unix time

so it means UUIDv7 needs no conversion?

> Added handling for values before the Unix epoch and for UUID
versions that do not store time.

Clarify on this. Perhaps, you mean your implementation carefully discards values before unix epoch, as well as unsupported UUIDs, and what is the result of that discard then.

> Added mysql-test coverage for known UUIDv1 and UUIDv7 values, UUID
typed input, NULL input, bad input with warnings, UUIDv4 returning
NULL, edge values, and TIMESTAMP(6) result metadata.

I'd reword it or remove it. mysql-test is a default test coverage so no need to mention it. It's like if you'd write "in SQL". Water is wet.

I think this bit can overall be read clearly from the test file.

> Used RFC 9562

Used hwo and how? If you mean you used it -- that's great, but how does it clarify the implementation details or the approach to your solution?

ParadoxV5 commented:
> i think i need the sign off for the cla

> Never heard of this.

It was not required yesterday.

varundeepsaini commented on plugin/type_uuid/mysql-test/type_uuid/func_uuid_timestamp.test:
```diff
@@ -0,0 +1,68 @@
+--echo #
+--echo # Native UUID type
+--echo #
+SELECT '019b6fdd-4937-7bb5-95c7-53363c6df927' AS uuid,
+       UUID_TIMESTAMP(CAST('019b6fdd-4937-7bb5-95c7-53363c6df927' AS UUID)) AS ts;
```
@FooBarrior

FooBarrior replied:
Right, thanks

grooverdan commented on plugin/type_uuid/mysql-test/type_uuid/func_uuid_timestamp.test:
```diff
@@ -0,0 +1,50 @@
+# MDEV-33710: UUID_TIMESTAMP() extracts timestamp from UUIDv1 and UUIDv7
```
```
--echo #
--echo # MDEV-33710: UUID_TIMESTAMP() extracts timestamp from UUIDv1 and UUIDv7
--echo #
```

grooverdan commented on include/my_sys.h:
```diff
@@ -1107,7 +1107,7 @@ int my_msync(int, void *, size_t, int);
 void my_uuid(uchar *guid);
 void my_uuid_end(void);
-
+int my_uuid_extract_ts(const char *uuid, my_time_t *seconds, ulong *usec);
```
space between a prototype list and the actual functions.

FooBarrior commented on plugin/type_uuid/item_uuidfunc.h:
```diff
@@ -115,4 +116,29 @@ class Item_func_uuid_v7: public Item_func_uuid_vx<UUIDv7>
+  LEX_CSTRING func_name_cstring() const override
+  {
+    static LEX_CSTRING name= {STRING_WITH_LEN("uuid_timestamp") };
+    return name;
```
This

FooBarrior reviewed (changes requested):
Now I can say this finally looks pretty good.

A few final (I hope) issues: see comment, and

* The commit message still refers to some "helper", without clarification of what it means:
> The helper returns failure

replace "The helper" with function name.

Also please make changes according to @grooverdan's requests.

varundeepsaini commented:
@FooBarrior 
I have sent the updated commit msg on zulip
